[confluent-local] Phase 2: add multi-broker option and broker management commands

MuweiHe opened the pull request:
Release Notes
-------------
<!--
If this PR introduces any user-facing changes, please document them below. Please delete any unused section titles and placeholders.
Please match the style of previous release notes: https://docs.confluent.io/confluent-cli/current/release-notes.html
-->

Breaking Changes
- PLACEHOLDER

New Features
- Add `--brokers` flag in `confluent local kafka start` to allow starting a multi-broker Confluent Local instance
- Add `confluent local kafka broker` command
- Add `confluent local kafka cluster configuration` command

Bug Fixes
- PLACEHOLDER

Checklist
---------
1. [CRUCIAL] Is the change for features that are already live in prod?
   * yes: ok

What
----
<!--
Briefly describe **what** you have changed and **why**.
Optionally include your implementation strategy.
-->
When running `confluent local kafka start`, can add a `multi-broker` flag to start the cluster as 2 brokers. Each broker runs on its own container. 
Added `broker list/describe/update` commands.

References
----------
<!--
Copy and paste links to tickets, related PRs, GitHub issues, etc.
-->

Test & Review
-------------
<!--
Has it been tested? How?
Copy and paste any instructions or steps that can save the reviewer time.
-->

Open Questions / Follow-ups
---------------------------
<!--
Optional: Anything open to discussion for the reviewer, out of scope of this PR, or follow-ups.
-->
In discussion: do we add a prompt for switching from single-node to 2-node when user runs `local start` and there's already an instance running.
Descriptions of `confluent local` should all change. It's not `single-node` necessarily anymore.
How to we help users to distinguish the 2 sets of old/new confluent local commands? how do we let them know that `confluent local broker` doesn't work with the old `services`?

davetroiano commented on internal/cmd/local/command_kafka_start.go:
```diff
@@ -127,46 +127,109 @@ func (c *Command) kafkaStart(cmd *cobra.Command, args []string) error {
 		Image:    dockerImageName,
-		Hostname: "broker",
-		Cmd:      strslice.StrSlice{"bash", "-c", "'/etc/confluent/docker/run'"},
+		Hostname: "broker1",
```
suggest hyphenating `broker-<ID>`.  reads a little more clearly. might be more consistent with other examples too -- I doubt we're consistent everywhere, but a couple of public multibroker examples that I checked do hyphenate (MRC [example](https://github.com/confluentinc/examples/blob/7.4.0-post/multiregion/docker-compose.yml#L67), kafka-images KRaft [example](https://github.com/confluentinc/kafka-images/blob/master/examples/confluent-server-kraft/docker-compose.yml#L20))

MuweiHe replied:
agree. This is also being used as the name of the containers too, so I'm thinking something more explicit like `confluent-local-broker-1`

davetroiano commented on internal/cmd/local/command_kafka_start.go:
```diff
@@ -227,59 +290,67 @@ func (c *Command) prepareAndSaveLocalPorts(cmd *cobra.Command, isTest bool) erro
+		fmt.Sprintf("KAFKA_ADVERTISED_LISTENERS=PLAINTEXT://broker%d:%s,PLAINTEXT_HOST://localhost:%s", brokerId, ports.BrokerPorts[idx], ports.PlaintextPorts[idx]),
 		"KAFKA_OFFSETS_TOPIC_REPLICATION_FACTOR=1",
 		"KAFKA_GROUP_INITIAL_REBALANCE_DELAY_MS=0",
 		"KAFKA_TRANSACTION_STATE_LOG_MIN_ISR=1",
 		"KAFKA_TRANSACTION_STATE_LOG_REPLICATION_FACTOR=1",
 		"KAFKA_PROCESS_ROLES=broker,controller",
-		"KAFKA_NODE_ID=1",
-		fmt.Sprintf("KAFKA_CONTROLLER_QUORUM_VOTERS=1@broker:%s", ports.ControllerPort),
-		fmt.Sprintf("KAFKA_LISTENERS=PLAINTEXT://broker:%s,CONTROLLER://broker:%s,PLAINTEXT_HOST://0.0.0.0:%s", ports.BrokerPort, ports.ControllerPort, ports.PlaintextPort),
+		fmt.Sprintf("KAFKA_NODE_ID=%d", brokerId),
+		fmt.Sprintf("KAFKA_CONTROLLER_QUORUM_VOTERS=1@broker1:%s,2@broker2:%s", ports.ControllerPorts[0], ports.ControllerPorts[1]),
+		fmt.Sprintf("KAFKA_LISTENERS=PLAINTEXT://broker%d:%s,CONTROLLER://broker%d:%s,PLAINTEXT_HOST://0.0.0.0:%s", brokerId, ports.BrokerPorts[idx], brokerId, ports.ControllerPorts[idx], ports.PlaintextPorts[idx]),
```
update these if broker name gets hyphenated

davetroiano commented on internal/cmd/local/command_kafka_start.go:
```diff
@@ -127,46 +127,109 @@ func (c *Command) kafkaStart(cmd *cobra.Command, args []string) error {
-	log.CliLogger.Tracef("Create confluent-local container success")
 
-	err = dockerClient.ContainerStart(context.Background(), createResp.ID, types.ContainerStartOptions{})
+	multiBroker, err := cmd.Flags().GetBool("multi-broker")
```
I think that we should go for configurable N brokers (1 <= N <= 5 or 10 or something) rather than a flag that gives you 2. I suspect users will want to play with other numbers, with 2 and 3 (common replication factor 3), maybe 4 being the common cases

davetroiano commented on internal/cmd/local/command_kafka_start.go:
```diff
@@ -127,46 +127,109 @@ func (c *Command) kafkaStart(cmd *cobra.Command, args []string) error {
-	output.Printf("Started Confluent Local container %v.\nTo continue your Confluent Local experience, run `confluent local kafka topic create test` and `confluent local kafka topic produce test`.\n", getShortenedContainerId(createResp.ID))
+		output.Printf("Started Confluent Local containers %v, %v.\nTo continue your Confluent Local experience, run `confluent local kafka topic create test` and `confluent local kafka topic produce test`.\n", getShortenedContainerId(createResp1.ID), getShortenedContainerId(createResp2.ID))
+	} else {
+		output.Printf("Started Confluent Local containers %v.\nTo continue your Confluent Local experience, run `confluent local kafka topic create test` and `confluent local kafka topic produce test`.\n", getShortenedContainerId(createResp1.ID))
```
This PR didnt introduce this, but in testing this I think that  the second sentence should be the last thing output, and maybe add a couple of newlines, e.g.:
```
Started Confluent Local containers da0e273845, 934727acfa.

+------------------+---------------+
| Broker Ports     | [58583 58584] |
| Controller Ports | [58585 58586] |
| Kafka Rest Ports | [8082 9092]   |
| Plaintext Ports  | [58581 58582] |
+------------------+---------------+

To continue your Confluent Local experience, run `confluent local kafka topic create test` and `confluent local kafka topic produce test`.
```

MuweiHe replied:
changed to only display krest port and plaintext ports as others are internal.
used comma-delimited format

davetroiano commented on internal/pkg/config/local_ports.go:
```diff
@@ -1,8 +1,8 @@
-	PlaintextPort  string `human:"Plaintext Port" json:"plaintext_port"`
+	BrokerPorts     []string `human:"Broker Ports" json:"broker_ports"`
+	ControllerPorts []string `human:"Controller Ports" json:"controller_ports"`
+	KafkaRestPorts  []string `human:"Kafka Rest Ports" json:"kafka_rest_ports"`
```
`Kafka REST Ports`?

davetroiano reviewed (commented):
suggest updating the stop command as part of this PR. currently it's this for multi-broker:
```
% ./confluent local kafka stop
Confluent Local has been stopped: removed container "934727acfa".
Confluent Local has been stopped: removed container "da0e273845".
```
suggest collecting the container IDs and printing one line:
```
% ./confluent local kafka stop
Confluent Local has been stopped: removed containers 934727acfa, da0e273845.
```

brianstrauch commented on internal/pkg/broker/broker_utils.go:
```diff
@@ -0,0 +1,130 @@
+package broker
```
Just calling this file `broker.go` is probably fine. Suffixing file names with `utils` makes it easy just to throw any function into it, and can quickly grow out of control (100+ lines of code)

brianstrauch commented on internal/cmd/kafka/command_broker_describe.go:
```diff
@@ -53,7 +38,7 @@ func (c *brokerCommand) newDescribeCommand() *cobra.Command {
 
 func (c *brokerCommand) describe(cmd *cobra.Command, args []string) error {
-	brokerId, all, err := checkAllOrBrokerIdSpecified(cmd, args)
+	brokerId, all, err := broker.CheckAllOrBrokerIdSpecified(cmd, args)
```
Since this is in the `broker` package, probably don't need to write "broker" in the function name
```suggestion
	brokerId, all, err := broker.CheckAllOrIdSpecified(cmd, args)
```

brianstrauch replied:
In the future, I want to get rid of the `--all` flag. For example,

```confluent kafka broker configuration list --all```

should probably be written as:

```confluent kafka cluster configuration list```

brianstrauch replied:
(Not sure how you're using `--all` in the Confluent Local commands, but maybe we should consider removing it pre-emptively)

brianstrauch commented on internal/cmd/local/command_broker.go:
```diff
@@ -0,0 +1,18 @@
+func (c *Command) newBrokerCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "broker",
+		Short: "Manage brokers in Confluent Local cluster",
```
```suggestion
		Short: "Manage Kafka brokers in Confluent Local.",
```

brianstrauch commented on internal/cmd/local/command_broker.go:
```diff
@@ -0,0 +1,18 @@
+		Short: "Manage brokers in Confluent Local cluster",
+	}
+
+	cmd.AddCommand(c.newBrokerListCommand())
+	cmd.AddCommand(c.newBrokerDescribeCommand())
```
```suggestion
	cmd.AddCommand(c.newBrokerDescribeCommand())
	cmd.AddCommand(c.newBrokerListCommand())
```

brianstrauch commented on internal/cmd/local/command_broker_describe.go:
```diff
@@ -0,0 +1,83 @@
+		),
+	}
+
+	cmd.Flags().Bool("all", false, "Get cluster-wide broker configurations (non-default values only).")
```
As I mentioned above, I would like to delete this and replace this with `confluent local kafka cluster configuration list` to match the existing CP command.

MuweiHe replied:
we have this flag in CP `confluent kafka broker describe --all` too. Are cluster configs and a cluster-wide broker config the same?

brianstrauch commented on internal/cmd/local/command_broker_update.go:
```diff
@@ -0,0 +1,103 @@
+func (c *Command) newBrokerUpdateCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "update [id]",
+		Short: "Update a Confluent local broker configurations.",
```
I believe the product "Confluent Local" is a proper noun that should be capitalized in all cases?
```suggestion
		Short: "Update Confluent Local broker configurations.",
```

brianstrauch commented on internal/cmd/local/command_kafka.go:
```diff
@@ -20,7 +21,7 @@ const (
 	cmd := &cobra.Command{
 		Use:   "kafka",
-		Short: "Manage a single-node instance of Apache Kafka.",
+		Short: "Manage a single-node or two-node instance of Apache Kafka.",
```
The distinction between a single- and multi-node cluster should probably just be part of the description of `--multi-broker`
```suggestion
		Short: "Manage a local instance of Apache Kafka.",
```

brianstrauch commented on internal/cmd/local/command_kafka_start.go:
```diff
@@ -37,14 +37,14 @@ type ImagePullResponse struct {
-
+	cmd.Flags().StringSlice("kafka-rest-ports", []string{"8082", "9092"}, "The port number for Kafka REST for brokers.")
+	cmd.Flags().StringSlice("plaintext-ports", nil, "The port number for plaintext producer and consumer clients for brokers. If not specified, a random free port will be used.")
+	cmd.Flags().Bool("multi-broker", false, "Start Confluent Local cluster with two brokers.")
```
```suggestion
	cmd.Flags().Bool("multi-broker", false, "Use two Kafka brokers instead of one.")
```

brianstrauch commented on internal/cmd/local/command_kafka_start.go:
```diff
@@ -127,46 +127,109 @@ func (c *Command) kafkaStart(cmd *cobra.Command, args []string) error {
+		return err
+	}
+	log.CliLogger.Tracef("Create confluent-local container for broker 1 success")
+	err = dockerClient.ContainerStart(context.Background(), createResp1.ID, types.ContainerStartOptions{})
 	if err != nil {
```
```suggestion
	if err := dockerClient.ContainerStart(context.Background(), createResp1.ID, types.ContainerStartOptions{}); err != nil {
```

brianstrauch commented on internal/cmd/local/command_kafka_start.go:
```diff
@@ -227,59 +290,67 @@ func (c *Command) prepareAndSaveLocalPorts(cmd *cobra.Command, isTest bool) erro
-	kafkaRestLn, err := net.Listen("tcp", ":"+c.Config.LocalPorts.KafkaRestPort)
-	if err != nil {
-		freePort, err := freeport.GetFreePort()
+	for idx := 0; idx < len(c.Config.LocalPorts.KafkaRestPorts); idx++ {
```
```suggestion
	for i, port := range c.Config.LocalPorts.KafkaRestPorts {
```

brianstrauch commented on internal/pkg/broker/broker_utils.go:
```diff
@@ -0,0 +1,130 @@
+	"github.com/spf13/cobra"
+)
+
+const AbbreviationLength = 25
```
Abbreviation length of what? The broker ID? The broker name?

MuweiHe replied:
when describing broker configs, the abbreved version of config name (e.g. ssl.trustmanager.algorith...

MuweiHe commented on internal/cmd/local/command_broker_list.go:
```diff
@@ -0,0 +1,55 @@
+	list := output.NewList(cmd)
+	for _, data := range brokersGetResp.Data {
+		broker := &broker.BrokerOut{
+			ClusterId: data.ClusterId,
```
for example cluster id would be just `4L6g3nShT-eMCtK--X86sw`. I'm ok with removing it since we only have one cluster.

MuweiHe commented:
@davetroiano @brianstrauch update the pr to allow configuring 1~4 brokers.

brianstrauch commented on internal/local/command_kafka_start.go:
```diff
@@ -27,6 +27,8 @@ import (
 	"github.com/confluentinc/cli/v3/pkg/output"
 )
 
+var defaultKafkaRestPorts = []string{"8082", "9092", "10012", "20012"}
```
9092 is typically used for the broker, not Kafka REST: https://docs.confluent.io/platform/current/installation/system-requirements.html#ports

brianstrauch commented on internal/local/command_kafka_start.go:
```diff
@@ -144,106 +149,61 @@ func (c *Command) kafkaStart(cmd *cobra.Command, args []string) error {
-					HostPort: ports.PlaintextPorts[0],
-				},
+	var containerIds []string
+	for idx := int32(0); idx < numOfBrokers; idx++ {
+		brokerId := idx + 1
```
```suggestion
	for idx := 0; idx < numOfBrokers; idx++ {
		brokerId := int32(idx + 1)
```

MuweiHe replied:
`numOfBrokers` is an int32

brianstrauch commented on internal/local/command_kafka_start.go:
```diff
@@ -144,106 +149,61 @@ func (c *Command) kafkaStart(cmd *cobra.Command, args []string) error {
 }
 
-func (c *Command) prepareAndSaveLocalPorts(cmd *cobra.Command, isTest bool) error {
+func (c *Command) prepareAndSaveLocalPorts(cmd *cobra.Command, numOfBrokers int32, isTest bool) error {
```
```suggestion
func (c *Command) prepareAndSaveLocalPorts(cmd *cobra.Command, numBrokers int32, isTest bool) error {
```

brianstrauch commented on internal/local/command_kafka_start.go:
```diff
@@ -256,16 +216,16 @@ func (c *Command) prepareAndSaveLocalPorts(cmd *cobra.Command, isTest bool) erro
 		}
 	} else {
-		freePorts, err := freeport.GetFreePorts(6)
+		freePorts, err := freeport.GetFreePorts(int(3 * numOfBrokers))
```
Why `* 3`? Does every broker have its own Kafka REST instance? That shouldn't be the case... right?

MuweiHe replied:
you're right. only one krest port is needed.

MuweiHe replied:
will update

MuweiHe replied:
krest is always 8082, but controller / broker / plaintext are 3 different ports

sgagniere commented on internal/local/command_kafka_start.go:
```diff
@@ -115,10 +117,13 @@ func (c *Command) kafkaStart(cmd *cobra.Command, args []string) error {
 	log.CliLogger.Tracef("Successfully pulled Confluent Local image")
 
-	if err := c.prepareAndSaveLocalPorts(cmd, c.Config.IsTest); err != nil {
+	numOfBrokers, err := cmd.Flags().GetInt32("brokers")
```
Based on the `// range: [1, 4]` comment, should that limit be enforced here? At a quick glance I didn't see anything that seems to stop users from choosing higher numbers.

MuweiHe replied:
that's right, no enforcement added yet!

brianstrauch commented on cmd/lint/main.go:
```diff
@@ -24,7 +24,7 @@ var commandRules = []linter.CommandRule{
 	linter.Filter(linter.RequireCapitalizeProperNouns("Short", properNouns), linter.ExcludeCommand("local current")),
 	linter.RequireEndWithPunctuation("Short", false),
-	linter.Filter(linter.RequireNotTitleCase("Short", properNouns)),
+	linter.Filter(linter.RequireNotTitleCase("Short", properNouns), linter.ExcludeCommandContains("local kafka broker")),
```
You shouldn't have to add this rule. Should we have added "Confluent Local" to the list of proper nouns?

brianstrauch commented on internal/kafka/command_broker_list.go:
```diff
@@ -48,7 +42,7 @@ func (c *brokerCommand) list(cmd *cobra.Command, _ []string) error {
 	list := output.NewList(cmd)
 	for _, data := range brokersGetResp.Data {
-		broker := &brokerOut{
+		broker := &broker.BrokerOut{
```
```suggestion
		broker := &broker.Out{
```

MuweiHe replied:
there're 2 types or `out` structs, config or broker

brianstrauch replied:
The package is called broker, so we can have: `broker.Out` and `broker.ConfigurationOut`

brianstrauch replied:
(`broker.BrokerOut` is a bit redundant, IMO)

brianstrauch commented on internal/local/command_kafka_broker_describe.go:
```diff
@@ -0,0 +1,83 @@
+	cmd := &cobra.Command{
+		Use:   "describe [id]",
+		Short: "Describe a local Kafka broker.",
+		Long:  "Describe cluster-wide or per-broker configuration values.",
```
I think we should align with the future state of the `broker` commands, that is:
1. This command should strictly be for printing `broker` configurations, not cluster-wide configurations (remove the `--all` flag)
2. Cluster-wide configurations should be printed with `confluent local kafka cluster configuration list`.

I was waiting for v4 to make this change, but seems like we have an opportunity to make it early here.

MuweiHe replied:
sounds good. Adding this change.

brianstrauch commented on internal/local/command_kafka_broker_describe.go:
```diff
@@ -0,0 +1,83 @@
+			return err
+		}
+		data = broker.ParseClusterConfigData(clusterConfig)
+	} else { // fetch individual broker configs
```
```suggestion
	} else {
		// fetch individual broker configs
```

brianstrauch commented on internal/local/command_kafka_broker_update.go:
```diff
@@ -0,0 +1,104 @@
+	"github.com/confluentinc/cli/v3/pkg/properties"
+)
+
+func (c *Command) newKafkaBrokerUpdateCommand() *cobra.Command {
```
These broker commands seem completely identical to the existing broker commands. Do you think we could call the main function of those commands directly?

MuweiHe replied:
how do i do this? kafka broker commands are of type `*pcmd.AuthenticatedCLICommand` 🤔

brianstrauch replied:
Sounds like we'll have to create the client outside of the function, and pass it in to a the shared function?

brianstrauch commented on internal/local/command_kafka_start.go:
```diff
@@ -37,14 +37,14 @@ type ImagePullResponse struct {
-
+	cmd.Flags().String("kafka-rest-port", "8082", "Kafka REST port number.")
+	cmd.Flags().StringSlice("plaintext-ports", nil, "A comma-separated list of port numbers for plaintext producer and consumer clients for brokers. If not specified, random free ports will be used.")
+	cmd.Flags().Int32("brokers", 1, "Number of brokers in the Confluent Local cluster.") // range: [1, 4]? do we enforce a range?
```
:eyes:

MuweiHe replied:
Haven't decided on the value we should set here. @davetroiano what's a proper range of number of brokers that users should be able to set?

brianstrauch commented on internal/kafka/command_broker_describe.go:
```diff
@@ -45,15 +30,15 @@ func (c *brokerCommand) newDescribeCommand() *cobra.Command {
 
 	cmd.Flags().Bool("all", false, "Get cluster-wide broker configurations (non-default values only).")
-	cmd.Flags().String("config-name", "", "Get a specific configuration value (pair with --all to see a cluster-wide configuration.")
+	cmd.Flags().String("config-name", "", "Get a specific configuration value (pair with --all to see a cluster-wide configuration).")
```
Good catch here. We can't use backticks here due to a limitation in Cobra, but regular quotes are better than nothing.
```suggestion
	cmd.Flags().String("config-name", "", `Get a specific configuration value (pair with "--all" to see a cluster-wide configuration).`)
```

brianstrauch commented on internal/local/command_kafka_broker_describe.go:
```diff
@@ -0,0 +1,70 @@
+		),
+	}
+
+	cmd.Flags().String("config-name", "", "Get a specific configuration value")
```
The linter should check for this, right?
```suggestion
	cmd.Flags().String("config-name", "", "Get a specific configuration value.")
```

brianstrauch commented on internal/local/command_kafka_broker_update.go:
```diff
@@ -0,0 +1,49 @@
+	}
+
+	pcmd.AddConfigFlag(cmd)
+	cmd.Flags().Bool("all", false, "Apply configuration update to all brokers in the cluster.")
```
I think we agreed to delete this in another comment, and use the `confluent kafka cluster configuration update` commands instead?

MuweiHe replied:
oops

brianstrauch commented on internal/local/command_kafka_cluster.go:
```diff
@@ -0,0 +1,16 @@
+	"github.com/spf13/cobra"
+)
+
+func (c *Command) newKafkaClusterCommand() *cobra.Command {
```
Any reason why `Command` is public?
```suggestion
func (c *command) newKafkaClusterCommand() *cobra.Command {
```

MuweiHe replied:
Good catch! It was not necessary

brianstrauch commented on internal/local/command_kafka_start.go:
```diff
@@ -127,79 +130,104 @@ func (c *Command) kafkaStart(cmd *cobra.Command, args []string) error {
+	_, err = dockerClient.NetworkCreate(
+		context.Background(),
+		"confluent-local-network",
+		types.NetworkCreate{
+			CheckDuplicate: true,
+			Driver:         "bridge",
 		},
```
Can you define this outside of the function call?

MuweiHe replied:
got it

brianstrauch reviewed (commented):
Looks like there are still comments from my previous review that haven't been addressed.

MuweiHe commented on internal/local/command_kafka_start.go:
```diff
@@ -127,79 +130,104 @@ func (c *Command) kafkaStart(cmd *cobra.Command, args []string) error {
-
-	output.ErrPrintln("To continue your Confluent Local experience, run `confluent local kafka topic create test` and `confluent local kafka topic produce test`.")
-
+	output.Printf("Started Confluent Local containers %s.\n", strings.Join(containerIds, ","))
```
Started Confluent Local containers "152f97f38c", "564f151276", "5f5ddca7d1",  "8c704f07df".

brianstrauch commented on internal/kafka/command_topic.go:
```diff
@@ -202,7 +202,7 @@ func ProduceToTopic(cmd *cobra.Command, keyMetaInfo []byte, valueMetaInfo []byte
 			break
 		}
-
+		fmt.Println("meta", keyMetaInfo, valueMetaInfo)
```
:eyes:

brianstrauch commented on internal/local/command_kafka_broker_describe.go:
```diff
@@ -0,0 +1,42 @@
+
+func (c *command) newKafkaBrokerDescribeCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "describe [id]",
+		Short: "Describe a local Kafka broker.",
+		Long:  "Describe per-broker configuration values.",
+		Args:  cobra.MaximumNArgs(1),
```
Should be required now that `--all` is removed.
```suggestion
		Use:   "describe <id>",
		Short: "Describe a local Kafka broker.",
		Long:  "Describe per-broker configuration values.",
		Args:  cobra.ExactArgs(1),
```

brianstrauch commented on internal/local/command_kafka_start.go:
```diff
@@ -127,79 +141,106 @@ func (c *Command) kafkaStart(cmd *cobra.Command, args []string) error {
+	natPlaintextPorts := getNatPlaintextPorts(ports, numBrokers)
+	containerStartCmd := strslice.StrSlice{"bash", "-c", "'/etc/confluent/docker/run'"}
+
+	_, err = dockerClient.NetworkCreate(
+		context.Background(),
+		"confluent-local-network",
+		networkCreate,
+	)
```
```suggestion
	_, err = dockerClient.NetworkCreate(context.Background(), "confluent-local-network", networkCreate)
```

brianstrauch commented on internal/local/command_kafka_start.go:
```diff
@@ -127,79 +141,106 @@ func (c *Command) kafkaStart(cmd *cobra.Command, args []string) error {
+			Image:    dockerImageName,
+			Hostname: fmt.Sprintf("confluent-local-broker-%d", brokerId),
+			Cmd:      containerStartCmd,
+			ExposedPorts: nat.PortSet{
+				natPlaintextPorts[idx]: struct{}{},
+			},
```
```suggestion
			ExposedPorts: nat.PortSet{natPlaintextPorts[idx]: struct{}{}},
```

brianstrauch commented on internal/local/command_kafka_start.go:
```diff
@@ -210,12 +251,12 @@ func (c *Command) prepareAndSaveLocalPorts(cmd *cobra.Command, isTest bool) erro
 		}
-		if plaintextPort != "" {
-			c.Config.LocalPorts.PlaintextPort = plaintextPort
+		if len(plaintextPorts) != 0 {
```
```suggestion
		if len(plaintextPorts) > 0 {
```

brianstrauch commented on internal/local/command_kafka_start.go:
```diff
@@ -230,60 +271,90 @@ func (c *Command) prepareAndSaveLocalPorts(cmd *cobra.Command, isTest bool) erro
+func getKafkaControllerQuorumVoters(ports *config.LocalPorts, numBrokers int32) string {
+	voters := fmt.Sprintf("KAFKA_CONTROLLER_QUORUM_VOTERS=1@confluent-local-broker-1:%s", ports.ControllerPorts[0])
+	for i := int32(1); i < numBrokers; i++ {
+		voters += fmt.Sprintf(",%v@confluent-local-broker-%v:%s", i+1, i+1, ports.ControllerPorts[i])
```
```suggestion
		voters += fmt.Sprintf(",%d@confluent-local-broker-%d:%s", i+1, i+1, ports.ControllerPorts[i])
```

brianstrauch commented on internal/local/command_kafka_topic.go:
```diff
@@ -61,3 +62,9 @@ func initKafkaRest(c *pcmd.CLICommand, cmd *cobra.Command) (*kafkarestv3.APIClie
+
+func (c *command) getPlaintextBootstrapServers() string {
+	servers := strings.Join(c.Config.LocalPorts.PlaintextPorts, ",:")
+	fmt.Println("bs:", ":"+servers)
```
:eyes:

MuweiHe commented on internal/local/command_kafka_start.go:
```diff
@@ -230,60 +271,90 @@ func (c *Command) prepareAndSaveLocalPorts(cmd *cobra.Command, isTest bool) erro
-		freePort, err := freeport.GetFreePort()
+	for idx, port := range c.Config.LocalPorts.PlaintextPorts {
+		brokerId := idx + 1
+		plaintextLn, err := net.Listen("tcp", ":"+port)
```
yes

brianstrauch replied:
```suggestion
		plaintextListener, err := net.Listen("tcp", ":"+port)
```

MuweiHe commented:
@brianstrauch thanks again for the review! fixed tests and lint, refactored broker functions, added `cluster configuration` commands, and removed println's :)

DABH reviewed (approved):
Hitting approve so I don’t get alerts about this but please work with the others for approval, don’t interpret my review as an actual approval

brianstrauch commented on internal/local/command_kafka_broker_update.go:
```diff
@@ -0,0 +1,44 @@
+func (c *command) newKafkaBrokerUpdateCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "update <id>",
+		Short: "Update local Kafka broker configurations.",
+		Long:  "Update per-broker configurations.",
```
`Long` is shorter than `Short` :thinking:

MuweiHe replied:
`"Update per-broker configurations for local Kafka brokers."`

brianstrauch commented on internal/local/command_kafka_cluster_configuration_update.go:
```diff
@@ -0,0 +1,77 @@
+	}
+
+	if output.GetFormat(cmd) == output.Human {
+		output.Printf("Updated the following broker configurations for cluster \"%s\":\n", clusterId)
```
Do we necessarily need this?

MuweiHe replied:
removed cluster id!

brianstrauch commented on internal/local/command_kafka_start.go:
```diff
@@ -25,30 +25,41 @@ import (
+	"github.com/confluentinc/cli/v3/pkg/utils"
 )
 
+var networkCreate = types.NetworkCreate{
```
This only gets used in one function. Let's define it there so it's not global in this package?

MuweiHe replied:
i took it out as per last comment said `define it outside the function`. I'm ok either way though

brianstrauch replied:
Hopefully that's not something I said 😅 Let's keep its scope as small as possible!

brianstrauch replied:
:eyes:

brianstrauch commented on internal/local/command_kafka_start.go:
```diff
@@ -127,79 +141,103 @@ func (c *Command) kafkaStart(cmd *cobra.Command, args []string) error {
-	output.ErrPrintln("To continue your Confluent Local experience, run `confluent local kafka topic create test` and `confluent local kafka topic produce test`.")
-
+	output.Printf("Started Confluent Local containers %s.\n", utils.ArrayToCommaDelimitedString(containerIds, ""))
+	output.Println("To continue your Confluent Local experience, run `confluent local kafka topic create test` and `confluent local kafka topic produce test`.")
```
```suggestion
	output.Println("To continue your Confluent Local experience, run `confluent local kafka topic create <topic>` and `confluent local kafka topic produce <topic>`.")
```

brianstrauch commented on internal/local/command_kafka_start.go:
```diff
@@ -230,60 +268,90 @@ func (c *Command) prepareAndSaveLocalPorts(cmd *cobra.Command, isTest bool) erro
+}
+
+func getKafkaControllerQuorumVoters(ports *config.LocalPorts, numBrokers int32) string {
+	voters := fmt.Sprintf("KAFKA_CONTROLLER_QUORUM_VOTERS=1@confluent-local-broker-1:%s", ports.ControllerPorts[0])
+	for i := int32(1); i < numBrokers; i++ {
+		voters += fmt.Sprintf(",%d@confluent-local-broker-%d:%s", i+1, i+1, ports.ControllerPorts[i])
+	}
```
Make a list of strings, and then use `strings.Join()`?

brianstrauch commented on internal/local/command_kafka_start.go:
```diff
@@ -230,60 +268,90 @@ func (c *Command) prepareAndSaveLocalPorts(cmd *cobra.Command, isTest bool) erro
+	return voters
+}
+
+func getKafkaRestBootstrapServers(ports *config.LocalPorts, numBrokers int32) string {
```
Seems like we could generalize this function and the one above?

MuweiHe replied:
they're using different ports and different ways to construct the string. it's clearer to keep them separate.

brianstrauch commented on internal/local/command_kafka_topic.go:
```diff
@@ -61,3 +62,8 @@ func initKafkaRest(c *pcmd.CLICommand, cmd *cobra.Command) (*kafkarestv3.APIClie
 }
+
+func (c *command) getPlaintextBootstrapServers() string {
+	servers := strings.Join(c.Config.LocalPorts.PlaintextPorts, ",:")
+	return ":" + servers
```
Can we loop through the array and add ":" to each of the elements, and then use `Join()`? This solution works, but seems really hacky.

MuweiHe commented on internal/local/command_kafka_start.go:
```diff
@@ -115,91 +133,121 @@ func (c *Command) kafkaStart(cmd *cobra.Command, args []string) error {
 	}
+	if numBrokers < 1 || numBrokers > 4 {
+		return errors.NewErrorWithSuggestions("Input number of brokers out of range.", "Choose the number of brokers between 1 and 4.") // wording?
+	}
```
@brianstrauch  I added a range check for between 1 and 4, but not sure about the error wording here. also is there other examples for enforcing the range of a int flag in our code?

brianstrauch replied:
How about the following?

> `--num-brokers` must be an integer between 1 and 4, inclusive.

sgagniere replied:
+1 to the flag wording.

As for the error, how about just reflecting the flag description, like: `number of brokers must be between 1 and 4, inclusive` and leave off the suggestion.

MuweiHe replied:
👍

brianstrauch commented on internal/local/command_kafka_start.go:
```diff
@@ -25,30 +25,48 @@ import (
+	"github.com/confluentinc/cli/v3/pkg/utils"
 )
 
+var (
```
```suggestion
const (
```

brianstrauch commented on internal/local/command_kafka_start.go:
```diff
@@ -25,30 +25,48 @@ import (
+}
+
+type portsOut struct {
+	KafkaRESTPort string `human:"Kafka Rest Port" json:"kafka_rest_port"`
```
```suggestion
	KafkaRestPort string `human:"Kafka REST Port" json:"kafka_rest_port"`
```

brianstrauch commented on internal/local/command_kafka_start.go:
```diff
@@ -25,30 +25,48 @@ import (
+
+type portsOut struct {
+	KafkaRESTPort string `human:"Kafka Rest Port" json:"kafka_rest_port"`
+	PlaintextPort string `human:"Plaintext Ports" json:"plaintext_ports"`
```
```suggestion
	PlaintextPorts string `human:"Plaintext Ports" json:"plaintext_ports"`
```

brianstrauch commented on internal/local/command_kafka_start.go:
```diff
@@ -25,30 +25,48 @@ import (
-
+	cmd.Flags().String("kafka-rest-port", "8082", "Kafka REST port number.")
+	cmd.Flags().StringSlice("plaintext-ports", nil, "A comma-separated list of port numbers for plaintext producer and consumer clients for brokers. If not specified, random free ports will be used.")
+	cmd.Flags().Int32("brokers", 1, "Number of brokers (between 1 and 4) in the Confluent Local cluster.")
```
```suggestion
	cmd.Flags().Int32("brokers", 1, "Number of brokers (between 1 and 4) in the Confluent Local Kafka cluster.")
```

brianstrauch commented on internal/local/command_kafka_start.go:
```diff
@@ -115,91 +133,121 @@ func (c *Command) kafkaStart(cmd *cobra.Command, args []string) error {
 	log.CliLogger.Tracef("Successfully pulled Confluent Local image")
 
-	if err := c.prepareAndSaveLocalPorts(cmd, c.Config.IsTest); err != nil {
+	numBrokers, err := cmd.Flags().GetInt32("brokers")
```
```suggestion
	brokers, err := cmd.Flags().GetInt32("brokers")
```

MuweiHe replied:
want to rename this flag as "num-brokers"

brianstrauch replied:
Up to you! Maybe check our other `int` and `int32` flags to see if there's a precendent?

brianstrauch commented on internal/local/command_kafka_start.go:
```diff
@@ -230,60 +278,90 @@ func (c *Command) prepareAndSaveLocalPorts(cmd *cobra.Command, isTest bool) erro
+
+func getNatPlaintextPorts(ports *config.LocalPorts, numBrokers int32) []nat.Port {
+	res := []nat.Port{}
+	for i := 0; i < int(numBrokers); i++ {
+		res = append(res, nat.Port(ports.PlaintextPorts[i]+"/tcp"))
+	}
```
```suggestion
	for _, port := range ports.PlaintextPorts {
		res = append(res, port+"/tcp")
	}
```

brianstrauch commented on test/fixtures/output/local/kafka/start-help.golden:
```diff
@@ -1,11 +1,12 @@
-      --plaintext-port string    The port number for plaintext producer and consumer clients. If not specified, a random free port will be used.
+      --kafka-rest-port string    Kafka REST port number. (default "8082")
+      --plaintext-ports strings   A comma-separated list of port numbers for plaintext producer and consumer clients for brokers. If not specified, random free ports will be used.
+      --brokers int32             Number of brokers (between 1 and 4, inclusive) in the Confluent Local Kafka cluster. (default 1)
```
`--num-brokers`?

MuweiHe replied:
other flags like `partitions` `cku` don't have that `--num` prefix so i decided to do it this way

brianstrauch replied:
Awesome, shorter is better! (I agree)

brianstrauch commented on test/fixtures/output/local/kafka/help.golden:
```diff
@@ -1,10 +1,12 @@
 
 Available Commands:
-  start       Start a single-node instance of Apache Kafka.
+  broker      Manage Kafka brokers in Confluent Local.
+  cluster     Manage Kafka cluster in Confluent Local.
```
Let's match the style of the other descriptions?
```suggestion
  broker      Manage local Kafka brokers.
  cluster     Manage local Kafka clusters.
```

brianstrauch commented on pkg/broker/broker_test.go:
```diff
@@ -1,43 +1,54 @@
 )
 
-func (suite *KafkaClusterTestSuite) TestBroker_checkAllOrBrokerIdSpecified() {
+type KafkaClusterTestSuite struct {
+	suite.Suite
+}
+
+func (suite *KafkaClusterTestSuite) TestBroker_CheckAllOrIdSpecified() {
```
This is a style of testing that we're trying to move away from. Can we use integration tests instead here?

MuweiHe replied:
we have broker tests for --all flag. I'll just remove this one.!

brianstrauch commented on internal/local/command_kafka_cluster.go:
```diff
@@ -7,7 +7,7 @@ import (
 	cmd := &cobra.Command{
 		Use:   "cluster",
-		Short: "Manage Kafka cluster in Confluent Local.",
+		Short: "Manage local Kafka cluster.",
```
... or if it makes more sense: "Manage the local Kafka cluster."
```suggestion
		Short: "Manage local Kafka clusters.",
```

MuweiHe replied:
"Manage the local Kafka cluster." sounds good

brianstrauch reviewed (approved):
thanks!!

MuweiHe commented:
> thanks!!

Ty!
